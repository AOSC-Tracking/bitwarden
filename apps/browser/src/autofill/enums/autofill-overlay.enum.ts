const AutofillOverlayElement = {
  Button: "autofill-inline-menu-button",
  List: "autofill-inline-menu-list",
} as const;

const AutofillOverlayPort = {
  Button: "autofill-inline-menu-button-port",
  ButtonMessageConnector: "autofill-inline-menu-button-message-connector",
  List: "autofill-inline-menu-list-port",
  ListMessageConnector: "autofill-inline-menu-list-message-connector",
} as const;

const RedirectFocusDirection = {
  Current: "current",
  Previous: "previous",
  Next: "next",
} as const;

export { AutofillOverlayElement, AutofillOverlayPort, RedirectFocusDirection };
