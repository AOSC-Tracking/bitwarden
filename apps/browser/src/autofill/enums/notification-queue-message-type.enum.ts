export const NotificationMessageType = {
  AddLogin: "add",
  ChangePassword: "change",
  UnlockVault: "unlock",
  RequestFilelessImport: "fileless-import",
} as const;

export type NotificationMessageTypes =
  (typeof NotificationMessageType)[keyof typeof NotificationMessageType];
