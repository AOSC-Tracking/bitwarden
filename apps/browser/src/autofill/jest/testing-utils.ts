function triggerTestFailure() {
  expect(true).toBe("Test has failed.");
}

const scheduler = typeof setImmediate === "function" ? setImmediate : setTimeout;
function flushPromises() {
  return new Promise(function (resolve) {
    scheduler(resolve);
  });
}

function postWindowMessage(data: any, origin = "https://localhost/") {
  globalThis.dispatchEvent(new MessageEvent("message", { data, origin }));
}

function sendExtensionRuntimeMessage(
  message: any,
  sender?: chrome.runtime.MessageSender,
  sendResponse?: CallableFunction
) {
  const onMessageMock = chrome.runtime.onMessage as any;
  if (onMessageMock.callListener) {
    onMessageMock.callListener(message, sender, sendResponse);
  }
}

function triggerWindowOnFocusedChangedEvent(windowId: number) {
  const onFocusChangedMock = chrome.windows.onFocusChanged as any;
  if (onFocusChangedMock.callListener) {
    onFocusChangedMock.callListener(windowId);
  }
}

function triggerTabOnActivatedEvent(activeInfo: chrome.tabs.TabActiveInfo) {
  const onActivatedMock = chrome.tabs.onActivated as any;
  if (onActivatedMock.callListener) {
    onActivatedMock.callListener(activeInfo);
  }
}

function triggerTabOnReplacedEvent(addedTabId: number, removedTabId: number) {
  const onReplacedMock = chrome.tabs.onReplaced as any;
  if (onReplacedMock.callListener) {
    onReplacedMock.callListener(addedTabId, removedTabId);
  }
}

function triggerTabOnUpdatedEvent(
  tabId: number,
  changeInfo: chrome.tabs.TabChangeInfo,
  tab: chrome.tabs.Tab
) {
  const onUpdatedMock = chrome.tabs.onUpdated as any;
  if (onUpdatedMock.callListener) {
    onUpdatedMock.callListener(tabId, changeInfo, tab);
  }
}

function triggerTabOnRemovedEvent(tabId: number, removeInfo: chrome.tabs.TabRemoveInfo) {
  const onRemovedMock = chrome.tabs.onRemoved as any;
  if (onRemovedMock.callListener) {
    onRemovedMock.callListener(tabId, removeInfo);
  }
}

export {
  triggerTestFailure,
  flushPromises,
  postWindowMessage,
  sendExtensionRuntimeMessage,
  triggerWindowOnFocusedChangedEvent,
  triggerTabOnActivatedEvent,
  triggerTabOnReplacedEvent,
  triggerTabOnUpdatedEvent,
  triggerTabOnRemovedEvent,
};
