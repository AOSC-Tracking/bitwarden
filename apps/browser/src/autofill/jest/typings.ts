type PortSpy = chrome.runtime.Port & {
  onDisconnect: { callListener: () => void };
  onMessage: { callListener: (message: any) => void };
};

export { PortSpy };
