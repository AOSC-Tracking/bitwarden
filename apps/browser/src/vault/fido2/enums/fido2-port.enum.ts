const Fido2Port = {
  InjectedScript: "fido2-injected-content-script-port",
} as const;

export { Fido2Port };
