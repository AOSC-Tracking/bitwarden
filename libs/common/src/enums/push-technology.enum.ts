export enum PushTechnology {
  SignalR = 0,
  WebPush = 1,
}
