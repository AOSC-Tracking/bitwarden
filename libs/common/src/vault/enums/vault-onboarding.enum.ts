const VaultOnboardingMessages = {
  HasBwInstalled: "hasBwInstalled",
} as const;

export { VaultOnboardingMessages };
