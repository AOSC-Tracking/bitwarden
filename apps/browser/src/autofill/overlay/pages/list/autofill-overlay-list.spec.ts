import { mock } from "jest-mock-extended";

import { AuthenticationStatus } from "@bitwarden/common/auth/enums/authentication-status";

import { createInitAutofillOverlayListMessageMock } from "../../../jest/autofill-mocks";
import { postWindowMessage } from "../../../jest/testing-utils";

import AutofillOverlayList from "./autofill-overlay-list";

describe("AutofillOverlayList", () => {
  globalThis.customElements.define("autofill-overlay-list", AutofillOverlayList);
  global.ResizeObserver = jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
  }));

  let autofillOverlayList: AutofillOverlayList;

  beforeEach(() => {
    document.body.innerHTML = `<autofill-overlay-list></autofill-overlay-list>`;
    autofillOverlayList = document.querySelector("autofill-overlay-list");
    autofillOverlayList["messageOrigin"] = "https://localhost/";
    jest.spyOn(globalThis.document, "createElement");
    jest.spyOn(globalThis.parent, "postMessage");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("initAutofillOverlayList", () => {
    describe("the locked overlay for an unauthenticated user", () => {
      beforeEach(() => {
        postWindowMessage(
          createInitAutofillOverlayListMessageMock({
            authStatus: AuthenticationStatus.Locked,
            cipherList: [],
          })
        );
      });

      it("creates the views for the locked overlay", () => {
        expect(autofillOverlayList["overlayListContainer"]).toMatchSnapshot();
      });

      it("allows the user to unlock the vault", () => {
        const unlockButton =
          autofillOverlayList["overlayListContainer"].querySelector("#unlock-button");

        unlockButton.dispatchEvent(new Event("click"));

        expect(globalThis.parent.postMessage).toHaveBeenCalledWith(
          { command: "unlockVault" },
          "https://localhost/"
        );
      });
    });
  });

  describe("global event listener handlers", () => {
    it("does not post a `checkAutofillOverlayButtonFocused` message to the parent if the overlay is currently focused", () => {
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(true);

      postWindowMessage({ command: "checkAutofillOverlayListFocused" });

      expect(globalThis.parent.postMessage).not.toHaveBeenCalled();
    });

    it("posts a `checkAutofillOverlayButtonFocused` message to the parent if the overlay is not currently focused", () => {
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(false);

      postWindowMessage({ command: "checkAutofillOverlayListFocused" });

      expect(globalThis.parent.postMessage).toHaveBeenCalledWith(
        { command: "checkAutofillOverlayButtonFocused" },
        "https://localhost/"
      );
    });

    describe("directing user focus into the overlay list", () => {
      it("focuses the unlock button element if the user is not authenticated", () => {
        postWindowMessage(
          createInitAutofillOverlayListMessageMock({
            authStatus: AuthenticationStatus.Locked,
            cipherList: [],
          })
        );
        const unlockButton =
          autofillOverlayList["overlayListContainer"].querySelector("#unlock-button");
        jest.spyOn(unlockButton as HTMLElement, "focus");

        postWindowMessage({ command: "focusOverlayList" });

        expect((unlockButton as HTMLElement).focus).toBeCalled();
      });
    });
  });

  describe("handleResizeObserver", () => {
    beforeEach(() => {
      postWindowMessage(createInitAutofillOverlayListMessageMock());
    });

    it("ignores resize entries whose target is not the overlay list", () => {
      const entries = [
        {
          target: mock<HTMLElement>(),
          contentRect: { height: 300 },
        },
      ];

      autofillOverlayList["handleResizeObserver"](entries as unknown as ResizeObserverEntry[]);

      expect(globalThis.parent.postMessage).not.toHaveBeenCalled();
    });

    it("posts a message to update the overlay list height if the list container is resized", () => {
      const entries = [
        {
          target: autofillOverlayList["overlayListContainer"],
          contentRect: { height: 300 },
        },
      ];

      autofillOverlayList["handleResizeObserver"](entries as unknown as ResizeObserverEntry[]);

      expect(globalThis.parent.postMessage).toHaveBeenCalledWith(
        { command: "updateAutofillOverlayListHeight", styles: { height: "300px" } },
        "https://localhost/"
      );
    });
  });
});
