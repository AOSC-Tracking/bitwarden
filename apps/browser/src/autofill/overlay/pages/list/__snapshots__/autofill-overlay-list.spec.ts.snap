// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutofillOverlayList initAutofillOverlayList the locked overlay for an unauthenticated user creates the views for the locked overlay 1`] = `
<div
  aria-modal="true"
  class="overlay-list-container light"
  role="dialog"
>
  <div
    class="locked-overlay overlay-list-message"
    id="locked-overlay-description"
  >
    unlockYourAccount
  </div>
  <div
    class="overlay-list-button-container"
  >
    <button
      aria-label="unlockAccount, opensInANewWindow"
      class="unlock-button overlay-list-button"
      id="unlock-button"
      tabindex="-1"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          clip-path="url(#a)"
        >
          <path
            d="M8.799 11.633a.68.68 0 0 0-.639.422.695.695 0 0 0-.054.264.682.682 0 0 0 .374.6v1.13a.345.345 0 1 0 .693 0v-1.17a.68.68 0 0 0 .315-.56.695.695 0 0 0-.204-.486.682.682 0 0 0-.485-.2Zm4.554-4.657h-7.11a.438.438 0 0 1-.406-.26A3.81 3.81 0 0 1 5.584 4.3c.112-.435.312-.842.588-1.195A3.196 3.196 0 0 1 7.19 2.25a3.468 3.468 0 0 1 3.225-.059A3.62 3.62 0 0 1 11.94 3.71l.327.59a.502.502 0 1 0 .885-.483l-.307-.552a4.689 4.689 0 0 0-2.209-2.078 4.466 4.466 0 0 0-3.936.185A4.197 4.197 0 0 0 5.37 2.49a4.234 4.234 0 0 0-.768 1.565 4.714 4.714 0 0 0 .162 2.682.182.182 0 0 1-.085.22.173.173 0 0 1-.082.02h-.353a1.368 1.368 0 0 0-1.277.842c-.07.168-.107.348-.109.53v7.1a1.392 1.392 0 0 0 .412.974 1.352 1.352 0 0 0 .974.394h9.117c.363.001.711-.142.97-.4a1.39 1.39 0 0 0 .407-.972v-7.1a1.397 1.397 0 0 0-.414-.973 1.368 1.368 0 0 0-.972-.396Zm.37 8.469a.373.373 0 0 1-.11.26.364.364 0 0 1-.26.107H4.246a.366.366 0 0 1-.26-.107.374.374 0 0 1-.11-.261V8.349a.374.374 0 0 1 .11-.26.366.366 0 0 1 .26-.108h9.116a.366.366 0 0 1 .37.367l-.008 7.097Z"
            fill="#175DDC"
          />
        </g>
        <defs>
          <clippath
            id="a"
          >
            <path
              d="M.798.817h16v16h-16z"
              fill="#fff"
            />
          </clippath>
        </defs>
      </svg>
      unlockAccount
    </button>
  </div>
</div>
`;
